### PR TITLE
Add HP and Hamilton filter

### PR DESCRIPTION
**Pull request by Shunsuke-Hori**
This PR adds "Hamilton filter" from [_Why You Should Never Use the Hodrick-Prescott Filter_](http://econweb.ucsd.edu/~jhamilto/hp.pdf)

Proposition 4 in the paper shows that the error term v_{t+h} of the regression

y_{t+h} = beta_0 + beta_1 y_{t} + ... + beta_p y{t-p+1} + v_{t+h}

gives cyclical component.

If `*arg` (p in the paper) is not passed, the true process is assumed to be a random walk. Then, we no longer need to do regression and we just take the difference.

Example:
~~~
df[:y_c], df[:y_t] = hamilton_filter(df[:y], 8, 4) # apply to data `y` in DataArray
df[:y_c], df[:y_t] = hamilton_filter(y, 8, 4) # apply to data `y` in `Vector`

# random walk
df[:y_c], df[:y_t] = hamilton_filter(df[:y], 8) # apply to data `y` in DataArray
df[:y_c], df[:y_t] = hamilton_filter(y, 8) # apply to data `y` in `Vector`
~~~

### CODE REVIEW

**Comment by Shunsuke-Hori** [comment]
I have no idea why the test is failing. It is passing the local test by `Pkg.test("QuantEcon")`

**Comment by sglyon** [comment]
Thanks for Submitting this. It looks great. 

Please let me know when you are ready for a full review. 

Thanks

**Comment by Simon-Mishricky** [comment]
I will be reviewing this PR

**Comment by Shunsuke-Hori** [comment]
Finally done! It seems `CSV.jl` didn't work correctly in travis...

@Simon-Mishricky Please review

**Comment by Shunsuke-Hori** [comment]
@sglyon This PR has been untouched for a while. @Simon-Mishricky must be busy. Could I ask you a review now? Or, should I wait a little more?

**Comment by sglyon** [comment]
Sorry about that --- given that we just pinged him again, let's give it one more week

**Comment by sglyon** [comment]
Hii @Shunsuke-Hori I'm sorry but I didn't have time to get to this this week and will be on holiday for the rest of the week.

If someone else on the team (like @jstac @oyamad @mmcky ) has a chance to review that would be great by me. Otherwise I'll get back to it next week.

Thanks for being patient.

**Comment by jstac** [comment]
This looks good and will be nice to have in the library.  It's well documented and I have no specific comments.  (The vectorized code for the HP filter is quite hard to follow, at least for me, but I guess it's fine.)

**Comment by sglyon** [comment]
Looks great, thanks.

i was wondering if we _need_ the test dependency on DataFrames. In this example could we get away with using a Dict instead of a DataFrame?

**Comment by Shunsuke-Hori** [comment]
Well, I thought people would want to use it with `DataFrames`, so I'm testing with it.

If it's unnecessary, we can drop it.

**Comment by sglyon** [comment]
Because you did a great job using AbstractArray we’ll be able to use arrays we unpack from a dict or a data frame. 

My vote would be to drop the requirement from the test suite just because I’m always a fan of reducing dependencies if possible

**Comment by sglyon** [comment]
Thank you — looks great. Merging now.
